src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -131,6 +164,56 @@ public static NpgsqlDataSource Create(string connectionString)
     public static NpgsqlDataSource Create(NpgsqlConnectionStringBuilder connectionStringBuilder)
         => Create(connectionStringBuilder.ToString());
 
+    /// <summary>
+    /// Manually sets the password to be used the next time a physical connection is opened.
+    /// Consider using <see cref="NpgsqlDataSourceBuilder.UsePeriodicPasswordProvider" /> instead.
+    /// </summary>
+    public string Password
+    {
+        set
+        {
+            if (_periodicPasswordProvider is not null)
+                throw new NotSupportedException(NpgsqlStrings.CannotSetBothPasswordProviderAndPassword);
+
+            _password = value;
+        }
+    }
+
+    internal async ValueTask<string?> GetPasswordAsync(bool async, CancellationToken cancellationToken = default)
+    {
+        // A periodic password provider is configured, but the first refresh hasn't completed yet (race condition).
+        // Wait until it completes.
+        if (_password is null && _passwordRefreshTask is not null)
```

Comment:
Nothing will actually set `_passwordRefreshTask` to null?

Reply:
Yeah, that's how it is - it seems harmless/not important. Setting to null would mean a bit of concurrency management which, well, who cares :)

Reply:
Sure, why not.